File: 01_Language/01_Functions/java/src/main/java/com/onevgo/functions/PregMatchAll.java
package com.onevgo.functions;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class PregMatchAll {
    public static boolean pregMatchAll(String pattern, String subject) {
        return pregMatchAll(Pattern.compile(pattern), subject);
    }

    public static boolean pregMatchAll(Pattern pattern, String subject) {
        Matcher matcher = pattern.matcher(subject);
        return matcher.find();
    }

    public static boolean pregMatchAll(String pattern, String subject, List<List<String>> matches) {
        return pregMatchAll(Pattern.compile(pattern), subject, matches);
    }

    public static boolean pregMatchAll(Pattern pattern, String subject, List<List<String>> matches) {
        Matcher matcher = pattern.matcher(subject);
        boolean b = false;
        while (matcher.find()) {
            for (int i = 0, j = matcher.groupCount(); i <= j; i++) {
                if (i > matches.size() - 1) {
                    matches.add(new ArrayList<>());
                }
                matches.get(i).add(matcher.group(i));
            }
            b = true;
        }
        return b;
    }

    public static void main(String[] args) {
        List<List<String>> matches = new ArrayList<>();
        System.out.println(pregMatchAll("\\(?(?:(\\d{3})-)?\\)?\\d{3}-\\d{4}", "Call 555-1212 or 1-800-555-1212", matches));
        System.out.println(matches);
    }
}

File: 01_Language/01_Functions/go/src/functions/preg_match_all.go
package functions

import (
	"regexp"
)

func PregMatchAll(pattern string, subject string, matches *[][]string) bool {
	compile, err := regexp.Compile(pattern)
	if err != nil {
		return false
	}
	matchList := compile.FindAllStringSubmatch(subject, -1)
	if len(matchList) == 0 {
		return false
	}
	if matches == nil {
		matches = new([][]string)
	}
	for _, items := range matchList {
		for k, v := range items {
			if k > len(*matches)-1 {
				*matches = append(*matches, []string{})
			}
			(*matches)[k] = append((*matches)[k], v)
		}
	}
	return true
}
